### PR TITLE
Allow nogil for most types of methods

### PR DESCRIPTION
**Pull request by jpfeuffer**
## Summary

This PR enables `nogil` support for most types of methods in autowrap-generated code. When a method is marked with `nogil` in the PXD file (e.g., `void process() nogil`), the generated wrapper will release Python's Global Interpreter Lock (GIL) before calling into C++, allowing true parallelism.

### Key Changes

- **ConversionProvider API change**: `input_conversion()` now returns a 4-tuple `(code, call_as, cleanup, decl)` where `decl` is a tuple `(declaration, call_expression)` for split declaration in nogil contexts
- **`call_method()` signature**: Added `with_cdef` parameter to control whether to include `cdef` (not allowed inside `nogil:` blocks)
- **CodeGenerator**: Updated to handle the new return values and generate proper nogil-compatible code
- **Cython requirement**: Updated to `>=3.1` for improved nogil support

### Updated Converters

All TypeConverter classes now support nogil:
- Primitive types (int, float, bool, char, etc.)
- String types (libcpp_string, const_char*, string_view)
- Container types (vector, map, set, list, deque, pair)
- C++17 types (optional, unordered_map, unordered_set)
- Wrapped class types
- Enum types

## nogil + OpenMP Thread Interaction Analysis

When a `nogil`-marked method (e.g., in pyOpenMS) internally uses OpenMP for parallelism:

### How It Works

1. **GIL is released** before entering the C++ code via `with nogil:` block
2. **OpenMP threads spawn** inside the C++ library
3. **All OpenMP threads execute** without competing for Python's GIL
4. **Threads complete**, control returns
5. **GIL is reacquired** when leaving the nogil block

### Safety Considerations

| Aspect | Behavior |
|--------|----------|
| **GIL during OpenMP** | Released - OpenMP threads run freely |
| **Thread safety** | Safe as long as C++ code doesn't call back into Python |
| **Python callbacks** | Would need `with gil:` to reacquire GIL first |
| **Other Python threads** | Can run concurrently while OpenMP works |

### Example Flow

```
Python: call pyopenms.method()
  └─ Cython: "with nogil:" releases GIL
       └─ C++ code runs
            └─ #pragma omp parallel spawns threads
                 └─ All threads work without GIL contention
            └─ OpenMP threads complete
       └─ Returns to Cython
  └─ GIL reacquired
Python: continues
```

### Bottom Line

**This is safe and beneficial** for libraries like OpenMS that use OpenMP internally:
- The GIL doesn't block OpenMP threads
- OpenMS's OpenMP code is pure C++ (no Python callbacks)
- Other Python threads can run while the C++ library processes data
- True multi-core parallelism is achieved

## Test Plan

- [x] All existing tests pass (165 passed, 1 skipped due to environment)
- [x] C++17 STL containers test passes (optional, string_view, unordered_map, unordered_set)
- [x] GIL testing with `gil_testing.hpp` verifies GIL is properly released

## References

- [Cython Parallelism Guide](https://cython.readthedocs.io/en/latest/src/userguide/parallelism.html)
- [Cython and the GIL](https://cython.readthedocs.io/en/latest/src/userguide/nogil.html)

### CODE REVIEW

**Comment by jpfeuffer** [comment]
Will fix tests because of newer typing syntax

**Review by poshul — COMMENTED** [review]
This looks good to me. Any idea what sort of performance improvements we are likely to see?

**Comment by jpfeuffer** [comment]
Probably none out of the box.
But it allows to use our functions in a multi threaded python context e.g. via the python threading module.

This brings some responsibilities for the user with it but you can decide which functions are safe to use.

**Review by timosachsenberg — APPROVED** [review]
Not super confident that I understand all parts. Would be great if in general methods would contain a brief comment e.g.the new _split one etc.

**Comment by jpfeuffer** [comment]
This is a breaking change. Be prepared to fix OpenMS if you merge this.

**Comment by poshul** [comment]
Given that we already need to do some fixes for OpenMS to get it to work with the current develop version of autowrap, I'm good merging this in and planning to have the core group sit down and fix at some point in the future.

**Comment by timosachsenberg** [comment]
what needs to be done to work with the current develop version of autowrap otherwise?

**Comment by timosachsenberg** [comment]
and what would need to be fixed?

**Comment by poshul** [comment]
> and what would need to be fixed?

https://github.com/OpenMS/autowrap/issues/186
